### PR TITLE
feat(ios, android): FLUID size support

### LINKED ISSUE #120: Extend AdSize support to handle size=FLUID
Extend AdSize support to handle size=FLUID

_Originally posted by @mikehardy in https://github.com/invertase/react-native-google-mobile-ads/pull/105#discussion_r858917392_

### LINKED ISSUE #213: [🐛] Fluid size not correctly
The banner size 'FLUID' does not work correctly. It does not take the parent width, it takes the full device/screen with.
I have a container around the ad with padding left and right. But the banner does not fit inside this container. Instead it goes outside the right side of the container.

```javascript
<View style={styles.container}>
      <GAMBannerAd
        unitId={__DEV__ ? TestIds.GAM_NATIVE : adUnitId}
        sizes={[BannerAdSize.FLUID]}
        requestOptions={{
          requestNonPersonalizedAdsOnly: !allowTracking,
        }}
      />
</View>
```

### PR DESCRIPTION
**Pull request by Gregoirevda**
### Description

Support for FLUID size on Android and fix for iOS FLUID size.

Android: FLUID size wasn't supported yet on Android. Fixes #120 
iOS: FLUID size was taking full device width. Fixes #213 

### Checklist

- I read the [Contributor Guide](https://github.com/invertase/react-native-google-mobile-ads/blob/main/CONTRIBUTING.md)
  and followed the process outlined there for submitting PRs.
  - [x] Yes
- My change supports the following platforms;
  - [x] `Android`
  - [x] `iOS`
- My change includes tests;
  - [ ] `e2e` tests added or updated in `__tests__e2e__`
  - [ ] `jest` tests added or updated in `__tests__`
- [ ] I have updated TypeScript types that are affected by my change.
- This is a breaking change;
  - [ ] Yes
  - [x] No

### Test Plan

---
Tested on various sizes including FLUID on Android and iOS, with different media (video/poll/creatives).

Think `react-native-google-mobile-ads` is great? Please consider supporting the project with any of the below:

- 👉 Star this repo on GitHub ⭐️
- 👉 Follow [`Invertase`](https://twitter.com/invertaseio) on Twitter

### CODE REVIEW

File: android/src/main/java/io/invertase/googlemobileads/ReactNativeGoogleMobileAdsBannerAdViewManager.java [hunk 1]
```diff
@@ -123,7 +126,7 @@ public void setSizes(ReactViewGroup reactViewGroup, ReadableArray value) {
       }
     }
 
-    if (sizeList.size() > 0) {
+    if (sizeList.size() > 0 && !sizeList.contains(AdSize.FLUID)) {
```
Comment (Gregoirevda) [inline android/src/main/java/io/invertase/googlemobileads/ReactNativeGoogleMobileAdsBannerAdViewManager.java:129]: No need to send dimensions to RN as RN View is set as `flex: 1` and the native view can now layout itself

File: android/src/main/java/io/invertase/googlemobileads/LayoutWrapper.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import android.widget.FrameLayout;
+
+import com.facebook.react.uimanager.ThemedReactContext;
+import com.facebook.react.uimanager.UIManagerModule;
+import com.facebook.react.views.view.ReactViewGroup;
+import com.google.android.gms.ads.AdSize;
+import com.google.android.gms.ads.BaseAdView;
+
+public class LayoutWrapper extends FrameLayout {
+  private BaseAdView contentView;
```
Comment (Gregoirevda) [inline android/src/main/java/io/invertase/googlemobileads/LayoutWrapper.java:14]: the `ViewGroup` can only contain a single `BaseAdView`

File: android/src/main/java/io/invertase/googlemobileads/LayoutWrapper.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+  public LayoutWrapper(Context context) {
+    super(context);
+  }
+
+  @Override
+  public void requestLayout() {
+    super.requestLayout();
+    post(measureAndLayout);
```
Comment (Gregoirevda) [inline android/src/main/java/io/invertase/googlemobileads/LayoutWrapper.java:23]: We're sending our own `measure` and `layout`

File: android/src/main/java/io/invertase/googlemobileads/LayoutWrapper.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  }
+
+  private final Runnable measureAndLayout = () -> {
+
+    BaseAdView adView = getContentView();
+
+    if(adView != null) {
```
Comment (Gregoirevda) [inline android/src/main/java/io/invertase/googlemobileads/LayoutWrapper.java:30]: If there's an ad set, measure and layout based on ad dimensions

File: android/src/main/java/io/invertase/googlemobileads/LayoutWrapper.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+      int top = adView.getTop();
+      int width = adSize.getWidthInPixels(getContext());
+      int height = adSize.getHeightInPixels(getContext());
+
+      measure(
+        MeasureSpec.makeMeasureSpec(width, MeasureSpec.EXACTLY),
+        MeasureSpec.makeMeasureSpec(height, MeasureSpec.EXACTLY));
+      layout(left, top, left + width, top + height);
+    } else {
+      measure(
```
Comment (Gregoirevda) [inline android/src/main/java/io/invertase/googlemobileads/LayoutWrapper.java:42]: If no ad set (requestLayout called before it was initialized), layout and measure based on ViewGroup

File: android/src/main/java/io/invertase/googlemobileads/LayoutWrapper.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    }
+  };
+
+
+  public void setContentView(BaseAdView view) {
+    contentView = view;
+    addView(contentView);
+  }
+
+  public BaseAdView getContentView() {
```
Comment (Gregoirevda) [inline android/src/main/java/io/invertase/googlemobileads/LayoutWrapper.java:55]: Remove the need of private method in `ReactNativeGoogleMobileAdsBannerAdViewManager.java`

File: android/src/main/java/io/invertase/googlemobileads/ReactNativeGoogleMobileAdsBannerAdViewManager.java [hunk 7]
```diff
@@ -38,14 +39,16 @@
...
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
+
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
 
 public class ReactNativeGoogleMobileAdsBannerAdViewManager
-    extends SimpleViewManager<ReactViewGroup> {
+  extends SimpleViewManager<LayoutWrapper> {
```
Comment (Gregoirevda) [inline android/src/main/java/io/invertase/googlemobileads/ReactNativeGoogleMobileAdsBannerAdViewManager.java:51]: ReactViewGroup overrides the [requestLayout](https://github.com/facebook/react-native/blob/0.66-stable/ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewGroup.java#L169), so extending `FrameLayout` which can only have a single child `BaseAdView` and setting our `requestLayout`

File: android/src/main/java/io/invertase/googlemobileads/ReactNativeGoogleMobileAdsBannerAdViewManager.java [hunk 8]
```diff
@@ -164,85 +167,68 @@ private BaseAdView initAdView(ReactViewGroup reactViewGroup) {
...
-
-          @Override
-          public void onAdClosed() {
-            sendEvent(reactViewGroup, EVENT_AD_CLOSED, null);
-          }
-        });
+      new AdListener() {
+        @Override
+        public void onAdLoaded() {
+          reactViewGroup.requestLayout();
```
Comment (Gregoirevda) [inline android/src/main/java/io/invertase/googlemobileads/ReactNativeGoogleMobileAdsBannerAdViewManager.java:173]: requestLayout after ad loaded, no need to send dimensions when size is FLUID

**Comment by Gregoirevda** [comment]
You actually need to wrap your `<GAMBannerAd` in a view with `height` and `width` eg
```jsx
<View style={{minHeight: 100, width: '100%'}}>
 <GAMBannerAd unitId="" sizes="FLUID" />
</View>
```

**Comment by mikehardy** [comment]
Hey! :wave: thanks for posting this - I was out traveling but I'll be back at it next week / during work week and I'll check through this. I appreciate the patience :pray:

**Comment by Gregoirevda** [comment]
Hi @mikehardy thank you for the update. Enjoy your holidays

File: src/ads/BaseAd.tsx [hunk 9]
```diff
@@ -115,8 +115,7 @@ export const BaseAd = React.forwardRef<GoogleMobileAdsBannerView, GAMBannerAdPro
 
     const style = sizes.includes(BannerAdSize.FLUID)
       ? {
-          width: '100%',
-          height: dimensions[1],
+          flex: 1
```
Comment (mikehardy) [inline src/ads/BaseAd.tsx:118]: @caiqueex this PR may be interesting for you if you want to take a look and help review

Comment (caiqueex) [reply src/ads/BaseAd.tsx:118]: Hi Mike, I o the test with this PR code. the result was it:

Captura de Tela 2022-12-10 às 23 26 20

I dont know why,  but it has that black background color and content to the right side

Comment (caiqueex) [reply src/ads/BaseAd.tsx:118]: Captura de Tela 2022-12-10 às 23 31 22

@mikehardy

Comment (Gregoirevda) [reply src/ads/BaseAd.tsx:118]: TestId.BANNER has 320x50. Does the black area represent 320+? Could you make a native ad on GAM and give it a 100% width and fixed height?

**Comment by Gregoirevda** [comment]
@mikehardy Anything I can do to help this being shipped?

**Comment by mikehardy** [comment]
Hey - sorry for the delay - the code seems okay but I'd love to see a snippet added to the example app that exercises this - I don't see where we are using the fluid style anywhere in the example app and there's no way for me to see it's working without an addition there

**Comment by dylancom** [comment]
Do we actually need "FLUID" support? The Mobile Ads SDK docs seems to solely focus on using the new anchored & inline adaptive banner ads. The problem of our current implementation is that an inline adaptive banner takes the screen's width instead of for e.g. the container views width.
https://developers.google.com/admob/ios/banner

**Comment by dylancom** [comment]
I made a PR that supports ads with custom width + auto height (using the new inline adaptive banner format). Would this make the FLUID size obsolete? https://github.com/invertase/react-native-google-mobile-ads/pull/296

**Comment by Gregoirevda** [comment]
This is to support native styles ads. They can only be fixed or fluid.

https://developers.google.com/ad-manager/mobile-ads-sdk/ios/native/styles#fluid_size

**Comment by dylancom** [comment]
Seems currently confusing for people like @caiqueex that actually need an adaptive banner with a custom width.

**Comment by MateusAndrade** [comment]
@Gregoirevda did you manage to render FLUID ads on Android/iOS with this patch?

**Comment by Gregoirevda** [comment]
@MateusAndrade yes

**Comment by MateusAndrade** [comment]
I see, I will test this patch then. I believe this is something worth it having in a new release. Should we open the pr again?

**Comment by mikehardy** [comment]
Hey @MateusAndrade 👋 😄 - happy to reopen this if it's useful for people. It's not a use case of mine so I'll need to know it is working, but we can merge it if it is. I fear there may be some conflicts as it has been closed a while - those will need resolution as well

**Comment by MateusAndrade** [comment]
> Hey @MateusAndrade 👋 😄 - happy to reopen this if it's useful for people. It's not a use case of mine so I'll need to know it is working, but we can merge it if it is. I fear there may be some conflicts as it has been closed a while - those will need resolution as well

Nice, I see that most of the changes introduced with this PR were using Java/Objective-C so I believe we will need to change this to Kotlin/Swift.

Btw, I was testing the iOS implementation but I didn't manage to make it. The AD was not showing, any clue @Gregoirevda ?

**Comment by Gregoirevda** [comment]
I've changed project meanwhile. Did you set a fixed width/height in the component? (windowWidth and fixed height for example) Can you share a code snippet

**Comment by MateusAndrade** [comment]
> I've changed project meanwhile. Did you set a fixed width/height in the component? (windowWidth and fixed height for example) Can you share a code snippet

Not really, I was testing using the sizes provided by library and indeed they work. But when I use `FLUID` on Android it doesn't render at all, while on iOS it renders but the AD doesn't fit in the slot, it ends overflowing it's parent. 

```jsx
    const sizes =
        Platform.OS === 'ios'
            ? ["FLUID"]
            : [BannerAdSize.MEDIUM_RECTANGLE, '336x280'];

    return (
                        <View>
                            <GAMBannerAd
                                unitId={AD_UNIT_ID}
                                sizes={sizes}
                                requestOptions={{
                                    customTargeting,
                                    publisherProvidedId,
                                }}
                            />
                        </View>
)
```

Where my `AD_UNIT` is either a real ad id or a test id. Should I use fixed dimensions for the AD container?

**Comment by Gregoirevda** [comment]
@MateusAndrade Not sure if it was on the `GAMBannerAd` or container `View`, but it was required.

[this](https://developers.google.com/ad-manager/mobile-ads-sdk/ios/native/styles#fluid_size) is what you currently have I think

**Comment by MateusAndrade** [comment]
> @MateusAndrade Not sure if it was on the `GAMBannerAd` or container `View`, but it was required.
> 
> [this](https://developers.google.com/ad-manager/mobile-ads-sdk/ios/native/styles#fluid_size) is what you currently have I think

I see, then we definitely got a bug here. Even if the parent component has fixed dimensions, the scaling to fit is not working. 😭

**Comment by Rajesh2github** [comment]
Hi when the FLUID ads size will work when we are planning to merge this PR

**Comment by mikehardy** [comment]
@Rajesh2github "we" ? Good question - I don't recall seeing you contribute to the PR. It looks like it has issues and if you need it you could work on the PR to fix the issues. We'd be happy to merge it at that point

**Pull request closed without merging**